### DISCUSSION TITLE
Using the pip-compile manager with private package registries

### DISCUSSION THREAD

**Discussion by mbudnek** [renovatebot/renovate#26854] [Suggest an Idea]
### Tell us more.

There is currently no way to use the pip-compile manager with private package registries when running the GitHub hosted version of Renovate.  This has previously been discussed in #21838 and https://github.com/renovatebot/renovate/discussions/24725#discussioncomment-8226384

I've got a package in a private packagecloud.io index that I'm trying to update via Renovate.  The initial pass at updating the package version in requirements.in works, since it delegates to the pip_requirements manager, but when shelling out to the `pip-compile` subprocess to generate the actual `requirements.txt` lockfile the the extra registry URL from the `packageRules` and the credentials from the `hostRules` configuration are lost.

Renovate needs a way to specify non-default registries to be passed to the pip-compile subprocess when updating the lockfile.  From looking into it, the two ways I can think of to accomplish this are:
1. Search the `packageRules` for rules with `matchDatasource: ["pypi"]` and find any `registryUrls` in them.
    - This would probably be fairly straightforward, but since those rules also apply to the pypi resolver it removes the ability of users to limit the scope of which packages that resolver resolves via each registry.  That may or may not actually be an issue though.
2. Search the `packageRules` for rules with `matchDatasource: ["pip-compile"]` and find any `registryUrls` in them.
    - pip-compile isn't really a datasource, so it's a bit of a misleading name, but this has the benefits of using an existing mechanism while also allowing the pypi resolver and the pip-compile manager to be configured independently.
3. Add a configuration option to the `"pip-compile"` config block.  This could just mean plumbing the the existing `registryUrls` config option down to the resolver.  This is the approach I proposed in #26853.
    - This also has the benefit of allowing the pypi resolver and the pip-compile manager to be configured independently.
4. Extract `--extra-index-url` flags from the requirements.in file and match them against `hostRules`
    - This matches what the pip_requirements manager already does, but it would require somehow getting those extracted URLs from the extract step to the update step.  I don't think that currently happens.  Also I don't think the extract step runs at all for lockfile maintenance, so the update step would probably have to re-parse the requirements.in file to some extent.  I also don't think this would work if the pip-compile manager were extended to work with the pip_setup manager for package extraction, which is something that it would ideally be able to support.

**Comment by not7cd** [comment]
Regarding 2: what are the cases when `pypi` resolver for `pip` must be configured differntly than `pypi` resolver for `pip-compile`?

**Reply by mbudnek** [reply]
I can imagine a case where someone has a repo with a requirements.in with all public dependencies that they use `pip-compile` to manage and a separate requirements-dev.txt with some private dev dependencies that they manage purely through `pip_requirements`.  In that case they may want to avoid allowing `pip-compile` to access their private repos to ensure no private packages inadvertently get added to the public requirements.
It's a pretty niche situation though.  I think coupling the configuration of the two would be OK, but there are edge cases where it could be undesirable.

**Reply by mbudnek** [reply]
Another option I just thought of would be to treat the lockfile itself more like a package, and key off the `matchPackagePatterns` in the `packageRules`.  That way you could individually configure which package registries `pip-compile` would use when generating each lockfile.  For example, in a situation similar to the above where you have private development/test dependencies, but want to have all public runtime dependencies you could do something like this:
```json
{
  "pip-compile": {
    "fileMatch": ["requirements-runtime\\.in", "requirements-dev\\.in"]
  },
  "packageRules": [
    {
      "matchPackagePatterns": ["mycompany-.*", "requirements-dev\\.txt" ],
      "registryUrls": [
        "https://pypi.python.org/pypi",
        "https://private.pypi.mydomain.com/simple"
      ]
    },
  ],
  "hostRules": [
    {
      "matchHost": "private.pypi.mydomain.com",
      "authType": "Basic",
      "username": "..."
      "password": "..."
    }
  ]
}
```
It has some of the same terminology issues as using `"matchDatasources": ["pip-compile"]`, since the lockfile isn't _really_ a package file, but this approach would be pretty flexible, and could avoid specifying the `registryUrls` multiple times when you _do_ want to share the same registries between the pypi resolver and the pip-compile manager.

**Reply by mbudnek** [reply]
Oh, wait, there's a `matchFileNames` setting in `packageRules` that would be perfect.  That makes way more sense than `matchPackagePatterns`.

**Reply by mbudnek** [reply]
Looking more into it, it doesn't look like the `packageRules` are currently available to managers in the `updateArtifacts` phase.  I could change that, but that would involve changing the manager interface, which @rarkins objected to in #26853

**Comment by rarkins** [comment]
I am still catching up on this topic. Initial questions:

1. These registry URLs aren't contained in the repo anywhere at all?
2. Doesn't Renovate need to know these registries in order to successfully look up the dependencies for newer versions?
3. These registries need to be included as part of the `pip-compile` command but they're not saved into the derived output file?

**Reply by mbudnek** [reply]
1. Not necessarily.  You _can_ include `--extra-index-url` lines in the .in file like I mentioned in (4) in the original post, but since there's no way to put any sort of credential placeholders in there it would mean leaving actual credentials laying around in Git.  Currently our developers either specify that flag on the command line or via an option in their global pip.conf file.  This is certainly an option though.  The `pip-compile` manager could re-parse the .in file looking for `--extra-index-url` lines and match them against `hostRules` to get credentials.  Developers would have to configure the credentials somehow.  This approach wouldn't work once the `pip-compile` manager gets support to work with the `pip_setup` manager though, since setup.py files won't have any sort of index URL.
2. The `pip-compile` manager delegates to the `pip_requirements` manager for parsing the .in file, and the `pypi` datasource for resolving direct dependencies.  The `pypi` datasource looks for `packageRules` with `matchDatasources: ["pypi"]` to look up newer dependency versions, but as I just discovered the `updateArtifacts` phase doesn't have access to `packageRules` to do the same thing.  There are also the edge cases I was discussing with @not7cd.
3. Again, since the registry URLs contain credentials (i.e. `https://username:password@pypi.my.domain/simple`) we can't write them out to the output file lest we leave developers' credentials laying around in Git (`pip-compile` provides the `--no-emit-index-url` flag to enable this). Even if we make sure not to have credentials in the `--extra-index-url` specified in the .in file, this could leak developer credentials into Git in the lockfile if they have such URLs configured globally and `--no-emit-index-url` isn't specified.

**Reply by rarkins** [reply]
For Renovate to lookup private registries two things need to happen:
1. Renovate needs to know which registry to look up. If the registry is in the repo somehow (e.g. .npmrc for npm) then it should be extracted automatically, otherwise you need to use packageRules with registryUrls to instruct Renovate where to look
2. Renovate needs to know how to authenticate with each registry host, and this is done using hostRules and usually token or username+password

If both the above are done then the lookups should succeed.

For updating artifacts, it means that Renovate needs to communicate the above to the third party tools. I thought that `registryUrls` are passed through if that's enough, and hostRules can be queried as done in most other managers.

As an alternative to passing registryUrls/hostRules, would supporting custom environment variables be easier for users to specify? We are planning to support that soon

**Reply by mbudnek** [reply]
Neither `regitryUrls` nor `packageRules` seem to be passed to `updateArtifacts`.  Adding `registryUrls` is one of the changes I made in #26853.
There is a `registryUrls` in the `updatedDeps` field of the config passed to `updateArtifacts`, but that's empty for lockfile maintenance.
I've added logic to extract the `--index-url` and `--extra-index-url` flags from the requirements.in file, which will be fine for my use-case.

Regarding the alternative of using environment variables, `pip-compile` does support `PIP_INDEX_URL` and `PIP_EXTRA_INDEX_URL` environment variables, so that's an option.  I believe that's how some people who self-host Renovate are working around this issue now.

**Reply by mbudnek** [reply]
I updated #26854 with my changes to pull the URLs from the input file.
